Add containerized retail workflow (#1574)
uyuni-project/uyuni-docs: modules/retail/pages/retail-install-setup.adoc

====
Ensure that SLES{nbsp}11 images are built on the SLES{nbsp}11 build host.
Building on the incorrect build host will cause your build to fail.

If you are building images for SLES{nbsp}11 using profiles from an HTTPS git repository that uses TLS 1.0 or greater, it will fail.
SLES{nbsp}11 does not support later versions of TLS.
You will need to clone the repository locally to use it for building.
====

== Branch identification and architecture topology

Before you configure the branch server, ensure you have decided on networking topology and you choose [systemitem]``branch id``.

> For the information about the possible network topologies, see xref:retail:retail-network-arch.adoc[].

Reviewer: ```suggestion
For information about the possible network topologies, see xref:retail:retail-network-arch.adoc[].
```

{productname} {smr} requires:

- branch system group for every branch server proxy, using [systemitem]``branch id`` as its name
- hardware type system group for every used hardware type, using [systemitem]``HWTYPE:`` prefix in its name

For more information about hardware type groups, see xref:retail:retail-deploy-terminals.adoc[].

[WARNING]
====
Missing mandatory system group will cause terminal bootstrap to fail.
====

{productname} {smr} also recognizes two optional groups for better organizations:

- [systemitem]``TERMINALS``
- [systemitem]``SERVERS``

You can create system groups using the {productname} {webui}.
Navigate to menu:Systems[System Groups] and click btn:[Create System Group].

For more information about system groups, see xref:reference:systems/system-groups.adoc[].

During terminal bootstrap terminal automatically joins:

- branch system group based on received [systemitem]``branch_id``. This will make branch group formulas available to the terminal.
- HWType group based on SMBios information received from terminal. This will make saltboot partitioning pillar available to the terminal.

> - [systemitem]``TERMINALS`` if this group exists

Reviewer: ```suggestion
- [systemitem]``TERMINALS`` if this group exists.
```

For more information about installing Ansible, see xref:administration:ansible-integration.adoc[].
====

== Configure Services for Saltboot

> Saltboot is mechanism which is used to deploy POS images to the terminals. Saltboot consists of saltboot enabled initrd (build as part of POS images) and saltboot salt states.

Reviewer: ```suggestion
Saltboot technology is used to deploy POS images to the terminals.
Saltboot consists of saltboot enabled initrd (build as part of POS images) and saltboot Salt states.
```

This section covers general information about generic Saltboot requirements.
For configuration examples, see xref:retail:example-configurations[].

=== Enable PXE network boot in the terminal network

> Saltboot is usually used in network boot environment. For this to work [systemitem]``DHCP`` service for the network terminal is connected to must have [systemitem]``PXE`` or sometimes called [systemitem]``BOOTP`` support enabled.

Reviewer: ```suggestion
Saltboot is usually used in network boot environment.
For this to work [systemitem]``DHCP`` service for the network terminal is connected to must have [systemitem]``PXE`` or sometimes called [systemitem]``BOOTP`` support enabled.
```

.Example of ISC DHCP server configuration with PXE booting enabled
====
  if substring (option vendor-class-identifier, 0, 10) = "HTTPClient" {
    option vendor-class-identifier "HTTPClient";
    filename "<FQDN of branch server proxy>/saltboot/shim.efi";
  }
  else {
    if option arch = 00:07 {
       filename "shim.efi";
       next-server <IP address of branch server proxy>;
    }
    else {
      filename "pxelinux.0";
      next-server <IP address of branch server proxy>;
    }
  }
====

Notice two important options, [systemitem]``next-server`` which is set to the branch server IP address and [systemitem]``filename`` set to the [systemitem]``pxelinux.0`` for BIOS based system and [systemitem]``shim.efi`` for UEFI systems with SecureBoot support.

=== Saltboot service discovery

> Saltboot requires some information where is the salt master and where to download image from. Saltboot tries multiple discoveries to obtain this information, described below.

Reviewer: ```suggestion
Saltboot requires some information where the Salt master is and from where to download the image.
Saltboot tries multiple discoveries to obtain this information, described below.
```

> For successful terminal deployment, both service discoveries must be successful. Depending on your architecture, choose what strategy works for you best.

Reviewer: ```suggestion
For successful terminal deployment, both service discoveries must be successful.
Depending on your architecture, choose what strategy works best for you.
```

==== Salt master discovery

> During saltboot initrd start, integrated salt minion needs to find branch server proxy to connect to. This discovery is trying following steps:

Reviewer: ```suggestion
During saltboot initrd start, integrated Salt client needs to find branch server proxy to connect to.
This discovery is trying following steps:
```

> - [systemitem]``MASTER`` kernel command line option is set, then this is used as salt master
> - resolve [systemitem]``salt`` CNAME, if successful then resolved value is used as salt master
> - use [systemitem]``salt`` as a salt master

Reviewer: ```suggestion
- [systemitem]``MASTER`` kernel command line option is set, then this is used as Salt master
- resolve [systemitem]``salt`` CNAME, if successful then resolved value is used as Salt master
- use [systemitem]``salt`` as a Salt master
```

> Once salt master is determined, salt minion configuration is generated and started.

Reviewer: ```suggestion
When Salt master is determined, Salt client configuration is generated and started.
```

[IMPORTANT]
====
Using fully qualified domain name in [systemitem]``MASTER`` or [systemitem]``salt`` CNAME is important.

If used fully qualified domain name is different from fully qualified domain name of branch server proxy known to {productname}, saltboot may work correctly, however proxy detection of terminal will not work.
====

==== Download server discovery

Before POS image is downloaded to the terminal, download server discovery is done to find where to download image from:

- [systemitem]``saltboot_download_server`` pillar is set for terminal, then its value is used
- [systemitem]``saltboot:download_server`` pillar is set for terminal, then its value is used
- resolve [systemitem]``ftp`` hostname

Value obtained by download server discovery is then used together with POS image pillar to fetch correct image from correct location.

=== Terminal partitioning and image selection

> Last piece for Saltboot is to provide partitioning for terminal. This is done individually for each hardware type of terminals.
> For more information about hardware types, see xref:retail:retail-deploy-terminals[]

Reviewer: ```suggestion
Last piece for Saltboot is to provide partitioning for terminal.
This is done individually for each hardware type of terminals.
For more information about hardware types, see xref:retail:retail-deploy-terminals[].
```

== Configure Services for Saltboot

Saltboot technology is used to deploy POS images to the terminals.

> Saltboot consists of saltboot enabled initrd (build as part of POS images) and saltboot salt states.

Reviewer: ```suggestion
Saltboot consists of saltboot enabled initrd (build as part of POS images) and saltboot Salt states.
```

=== Terminal partitioning and image selection

Last piece for Saltboot is to provide partitioning for terminal.
This is done individually for each hardware type of terminals.

> For more information about hardware types, see xref:retail:retail-deploy-terminals.adoc[]

Reviewer: ```suggestion
For more information about hardware types, see xref:retail:retail-deploy-terminals.adoc[].
```
